snuba/schemas.py

```diff
@@ -217,42 +183,149 @@ def get_time_series_query_schema_properties(default_granularity: int, default_wi
+                    'type': 'array',
+                    'items': {'type': 'number'},
+                    'minItems': 1,
+                },
+            ]
+        },
+    },
     # Need to select down to the project level for customer isolation and performance
     'required': ['project'],
+    'additionalProperties': False,
 }
 
-EVENTS_QUERY_SCHEMA['properties'].update({
-    **get_time_series_query_schema_properties(
+
+def get_time_series_extension_properties(default_granularity: int, default_window: timedelta):
+    return {
+        'type': 'object',
+        'properties': {
+            'from_date': {
+                'type': 'string',
+                'format': 'date-time',
+                'default': lambda: (datetime.utcnow().replace(microsecond=0) - default_window).isoformat()
+            },
+            'to_date': {
+                'type': 'string',
+                'format': 'date-time',
+                'default': lambda: datetime.utcnow().replace(microsecond=0).isoformat()
+            },
+            'granularity': {
+                'type': 'number',
+                'default': default_granularity,
+            },
+        },
+        'additionalProperties': False,
+    }
+
+
+@dataclass
+class Request:
+    query: Mapping[str, Any]
+    extensions: Mapping[str, Mapping[str, Any]]
+
+    def __post_init__(self):
+        self.body = ChainMap(self.query, *self.extensions.values())
+
+
+class RequestSchema:
+    def __init__(self, query_schema, extensions_schemas: Mapping[str, Any]):
+        self.__query_schema = query_schema
+        self.__extension_schemas = extensions_schemas
+
+        self.__composite_schema = {
+            'type': 'object',
+            'properties': {},
+            'required': [],
+            'definitions': {},
+            'additionalProperties': False,
+        }
+
+        for schema in itertools.chain([self.__query_schema], self.__extension_schemas.values()):
+            assert schema['type'] == 'object', 'subschema must be object'
+            assert schema['additionalProperties'] is False, 'subschema must not allow additional properties'
+            self.__composite_schema['required'].extend(schema.get('required', []))
+
+            for property_name, property_schema in schema['properties'].items():
+                assert property_name not in self.__composite_schema['properties'], 'subschema cannot redefine property'
+                self.__composite_schema['properties'][property_name] = property_schema
+
+            for definition_name, definition_schema in schema.get('definitions', {}).items():
+                assert definition_name not in self.__composite_schema['definitions'], 'subschema cannot redefine definition'
+                self.__composite_schema['definitions'][definition_name] = definition_schema
+
+        self.__composite_schema['required'] = set(self.__composite_schema['required'])
+
+    def validate(self, value) -> Request:
+        # XXX: Mutates input value!
```

Comment:
Is it worth making a copy just to stay on the safe side here? 
None is supposed to use value after this method anyway so if we throw that result away it is not a an issue. On the other hand we make things safer for the call site.

Reply:
Sure, I'll do that.

Reply:
I moved the copy down to `validate` (and renamed it `validate_jsonschema` to be more specific) in 70b6e8bd3e38c6e3f28308ff3a43cc8cc480487c, this seemed like a more appropriate place to actually make that change than here.